### PR TITLE
add custom middlewares option

### PR DESCRIPTION
**Pull request by zebox**
In this PR implemented custom middlewares option as mention #8. The options `CustomMiddlewares`  added direct to `Server` struct with type  `Middlewares` which was define too.

### CODE REVIEW

File: server.go
Comment (umputun) [inline server.go:70]: I think we should keep throttler, realio and recoverer in the default set of middlewares. If throttle is smth not tha obvious but both realip and recoverer are must have

**Review by umputun — CHANGES_REQUESTED** [review]
thx, a small suggestion. In addition readme wording is somewhat confusing but I'll rephrase it after the merge

File: server.go
Comment (umputun) [reply server.go:70]: i know the original ticket mentioned realip and recoverer as something user may pass, but I can't see any use case where user may want to skip these two. Well, maybe realip (not sure why, but anyway) but lack of recoverer will fail server on errors

**Comment by zebox** [comment]
If are you don't have remarks could you LGTM?

**Comment by umputun** [comment]
I don't see any problem with your PR on the code level. My only concern is "silently braking" nature of this change. I.e. consumer gets this version and the behavior changes behind the scene. This could be something unexpected and even potentially dangerous.  I would rather prefer a breaking change of some sort, very explicit one instead of the current implicit change of the logic.

I'm not sure what the best/right solution here. We can accept the risk, we can make not-compatible change, or we can make the old configuration default and allow "undefault" it somehow, or maybe something else.

**Comment by zebox** [comment]
I understand you fear and you are righ. May be stay all previous middlewares by default as you offer and mark they as deprecated for future release?

I need add custom middlewares only  and use main stream of you package. If you think solution which I offer dangerous I will use fork for my case.

**Comment by umputun** [comment]
I think the best we can do here is to break the current API intentionally. This is what I'd suggest:

- make Server constructible with NewServer and change all the fields to private
- the constructor will get a single parameter `API`
- everything else, including the custom list of middlewares passed in with optional func pattern, like [here](https://github.com/go-pkgz/lgr/blob/master/options.go)

This way we make the change breaking, so no confusions on silently changed functionality. Also, dealing with default/non-default will be much easier, for example one of options will be `WithLogger` and by default it will set nop logger. The same logic will be about all the other parts, i.e. WithThrottler, WithLimits and so on. Obviously custom middlewares will have `WithMiddlewares` option

what do you think?

**Comment by zebox** [comment]
I think it's good idea. But I'm thinking close this PR, because discussed changes not relevant with it.

I will add change with new options for server configure and will open new PR.

**Pull request closed without merging**
